src/sidebar/components/test/tag-editor-test.js

```diff
@@ -1,63 +1,257 @@
+  });
+
+  it('generates a datalist set equal to the array value returned from fakeTagsService.filter ', () => {
+    const wrapper = createComponent();
+    // simulate `keyup` to populate suggestions list
+    wrapper.find('input').simulate('keyup', { key: 'any' });
+
+    assert.equal(
+      wrapper
+        .find('datalist option')
+        .at(0)
+        .prop('value'),
+      'tag3'
+    );
+    assert.equal(
+      wrapper
+        .find('datalist option')
+        .at(1)
+        .prop('value'),
+      'tag4'
+    );
+  });
+
+  it('clears the suggestions when adding a new tag', () => {
+    const wrapper = createComponent();
+    wrapper.find('input').instance().value = 'tag3';
+    // simulate `keyup` to populate suggestions list
+    wrapper.find('input').simulate('keyup', { key: 'any' });
+    assert.equal(wrapper.find('datalist option').length, 2);
+    wrapper.find('input').simulate('keyup'); // no key supplied
+    assert.equal(wrapper.find('datalist option').length, 0);
+
+    assert.isTrue(
+      fakeTagsService.store.calledWith([
+        { text: 'tag1' },
+        { text: 'tag2' },
+        { text: 'tag3' },
+      ])
+    );
+  });
+
+  it('does not render duplicate suggestions', () => {
+    // `tag3` supplied in the `tagList` will be a duplicate value relative
+    // with the fakeTagsService.filter result above.
+    const wrapper = createComponent({
+      editMode: true,
+      tagList: ['tag1', 'tag2', 'tag3'],
     });
-    assert.deepEqual(element.ctrl.tagList, [{ text: 'foo' }, { text: 'bar' }]);
+    // simulate `keyup` to populate suggestions list
+    wrapper.find('input').simulate('keyup', { key: 'any' });
+    assert.equal(wrapper.find('datalist option').length, 1);
+    assert.equal(
+      wrapper
+        .find('datalist option')
+        .at(0)
+        .prop('value'),
+      'tag4'
+    );
   });
 
-  describe('when tags are changed', function() {
-    let element;
-    let onEditTags;
+  context('when adding tags', () => {
+    /**
+     * Helper function to assert that a tag was correctly added
+     */
+    const addTagsSuccess = (wrapper, tagList) => {
+      // saves the suggested tags to the service
+      assert.isTrue(
+        fakeTagsService.store.calledWith(tagList.map(tag => ({ text: tag })))
+      );
+      // called the onEditTags callback prop
+      assert.isTrue(fakeOnEditTags.calledWith({ tags: tagList }));
+      // clears out the suggestions
+      assert.equal(wrapper.find('datalist option').length, 0);
+      // assert the input value is cleared out
+      assert.equal(wrapper.find('input').instance().value, '');
+      // note: focus not tested
+    };
+    /**
+     * Helper function to assert that a tag was correctly not added
+     */
+    const addTagsFail = () => {
+      assert.isTrue(fakeTagsService.store.notCalled);
+      assert.isTrue(fakeOnEditTags.notCalled);
+    };
+
+    it('adds a tag from the input field', () => {
```

Comment:
I wonder if this test name could be expanded to indicate _when_ this adding of tag is expected to happen.